src/ieee8023.c

```diff
@@ -0,0 +1,502 @@
+
+	if (is_frag && !frag_id)
+		frag_id = uatomic_add_return(&wtp->fragment_id, 1);
+
+	debug("frag_id: %d", frag_id);
+
+	memset(chdr, 0, sizeof(chdr));
+
+	/* Version = 0, Type = 0, HLen = 8, RId = 1, WBID 802.11, 802.3 encoding */
+	((uint32_t *)chdr)[0] = htobe32((hlen << CAPWAP_HLEN_SHIFT) |
+					(rid << CAPWAP_RID_SHIFT) |
+					(CAPWAP_802_11_PAYLOAD << CAPWAP_WBID_SHIFT));
+
+	if (wbinfo != NULL && wbinfo_len != 0)
+		((uint32_t *)chdr)[0] |= CAPWAP_F_WSI;
+
+	if (is_frag) {
+		((uint32_t *)chdr)[0] |= CAPWAP_F_FRAG;
+		uatomic_inc(&w->send_fragments);
+		uatomic_inc(&wtp->send_fragments);
+	}
+	((uint32_t *)chdr)[1] = htobe32(frag_id << CAPWAP_FRAG_ID_SHIFT);
+
+	/* The message header contains parameters for sendmsg.    */
+	memset(&mh, 0, sizeof(mh));
+	mh.msg_name = (caddr_t)&wtp->addr;
+	mh.msg_namelen = sizeof(wtp->addr);
+	mh.msg_iov = iov;
+
+	iov[0].iov_base = chdr;
+	iov[0].iov_len = hlen * 4;
+
+	piov = 1;
+	if (wbinfo != NULL && wbinfo_len != 0) {
+		/* WSI */
+		iov[piov].iov_base = (unsigned char *)wbinfo;
+		iov[piov].iov_len = wbinfo_len;
+		piov++;
+	}
+
+	iov[piov].iov_base = (unsigned char *)buffer;
+	iov[piov].iov_len = frag_size;
+
+	mh.msg_iovlen = piov + 1;
+
+	ssize_t offs = 0;
+
+	for (; frag_count > 0; frag_count--) {
+		debug("Id: %d, Count: %d", frag_id, frag_count);
+		if (caa_unlikely(is_frag)) {
+			SET_CAPWAP_HEADER_FIELD(&((uint32_t *)chdr)[1], offs / 8, CAPWAP_FRAG_OFFS_MASK, CAPWAP_FRAG_OFFS_SHIFT);
+			if (frag_count == 1) {
+				((uint32_t *)chdr)[0] |= CAPWAP_F_LASTFRAG;
+				iov[piov].iov_len = len % frag_size;
+			}
+		}
+
+		/* FIXME: shortcat write, we do want to use NON-BLOCKING send here and
+		 *        switch to write_ev should it block....
+		 */
+		if ((r = sendmsg(w->capwap_fd, &mh, MSG_DONTWAIT)) < 0)
+			debug("sendmsg: %m");
+
+		debug("fwd tap sendmsg: %zd", r);
+
+		if (r > 0) {
+			uatomic_inc(&w->send_pkts);
+			uatomic_inc(&wtp->send_pkts);
+			uatomic_add(&w->send_bytes, r);
+			uatomic_add(&wtp->send_bytes, r);
+		}
+
+		if (caa_unlikely(((uint32_t *)chdr)[0] & CAPWAP_F_WSI)) {
+			/* don't send WSI on following fragements */
+			piov--;
+			mh.msg_iovlen--;
+
+			((uint32_t *)chdr)[0] &= ~CAPWAP_F_WSI;
+			iov[piov].iov_base = (unsigned char *)buffer;
+			iov[piov].iov_len = frag_size;
+		}
+
+		iov[piov].iov_base = (unsigned char *)iov[piov].iov_base + frag_size;
+		offs += frag_size;
+	}
+
+	return 1;
+}
+
+struct ether_header* fill_raw_udp_packet(void *data, uint16_t data_len,
+        struct in_addr *saddr, uint8_t *mac_shost,
+        struct in_addr *daddr, uint8_t *mac_dhost, uint16_t *send_len)
+{
+    uint8_t *datagram, *data_buf;
+    uint16_t ip_tot_len, tot_len;
+    uint32_t udp_csum;
+
+    ip_tot_len = data_len + sizeof(struct iphdr) + sizeof(struct udphdr);
+    tot_len = ip_tot_len + sizeof(struct ether_header);
+    *send_len = tot_len;
+
+    datagram = (uint8_t *) calloc(1, tot_len);
```

Comment:
This should work with calloc. It might be better to pass a iovec through to ieee8023_to_{sta,wtp}.

Reply:
You mean that i should use malloc instead calloc or just remove calloc?